rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,689 @@
+    *   Friendly to other OSS data formats, both in-memory and on disk (e.g.
+        Pandas)
+    *   Friendly to numpy / TF: many Arrow array types share the same memory
+        layout with numpy ndarrays and certain type of TF (composite) Tensors.
+*   TF neutral.
+    *   Leaves the possibility of supporting other ML libraries open.
+
+## Translation from Arrow to TF feedables
+
+The analogy to this is parsing tf.Examples into TF feedables -- extra
+information is needed in this translation because a
+[`Feature`](https://github.com/tensorflow/tensorflow/blob/abfba15cd9734cec7ecd3d0661b146fc251c842d/tensorflow/core/example/feature.proto#L76)
+can be converted to a Tensor, a SparseTensor or a
+[RaggedTensor](https://www.tensorflow.org/guide/ragged_tensor) depending on the
+[feature specs](https://github.com/tensorflow/tensorflow/blob/635e23a774936b5fe6fa3ef3cb6e54b55d93f324/tensorflow/python/ops/parsing_ops.py#L46-L49).
+Currently this extra information is implicitly contained in the pipeline schema
+(an instance of the
+[TFMD Schema](https://github.com/tensorflow/metadata/blob/master/tensorflow_metadata/proto/v0/schema.proto))
+proto.
+
+Similarly, an Arrow column can be translated to various TF feedables.
+[An extension to the pipeline schema](#tensorrepresentation) is proposed to for
+a user to express the intention for conversion.
+
+The conversion can be efficient (zero-copy) in certain cases. It is
+[discussed](#efficient-arrow-tensor-conversion) in the detailed design.
+
+## Standardized Inputs APIs
+
+We propose a set of APIs that TFX components will call, and need to be
+implemented for each of the supported combination of {physical, logical} format.
+
+```py
+class TFXIO(object):
+  """Abstract basic class of all Standardized TFX inputs API implementations."""
+  def __init__(
+      self, pipeline_env,
+      schema: Optional[tfmd.Schema]=None
+  ):
+    pass
+
+  @abc.abstractmethod
+  def BeamSource(self,
+                 projections: Optional[List[Text]]=None
+  ) -> beam.PTransform:
+    """Returns a beam PTransform that produces PCollection[pa.RecordBatch].
+
+    May NOT raise an error if the TFMD schema was not provided at construction time.
+
+    Args:
+      projections: if not None, only the specified subset of columns will be
+      read.
+    """
+
+  @abc.abstractmethod
+  def TensorAdapter(self) -> TensorAdapter:
+    """Returns a TensorAdapter that converts pa.RecordBatch to TF inputs.
+
+    May raise an error if the TFMD schema was not provided at construction time.
+    """
+
+  @abc.abstractmethod
+  def ArrowSchema(self) -> pyarrow.Schema:
+    """Returns the schema of the Arrow RecordBatch generated by BeamSource().
+
+    May raise an error if the TFMD schema was not provided at construction time.
+    """
+
+  @abc.abstractmethod
+  def TFDataset(self, ...) -> tf.data.Dataset:
+    """Returns a Dataset of TF inputs.
+
+    May raise an error if the TFMD schema was not provided at construction time.
```

Comment:
Looks like schema is strictly required for those methods, so they must raise (as opposed to may) an error if schema is absent?